torch_xla/distributed/spmd/debugging.py

```diff
@@ -0,0 +1,207 @@
+from collections.abc import Sequence
+import functools
+import string
+import sys
+from typing import Any, Callable, Optional, Union
+import weakref
+
+import numpy as np
+import torch
+import torch_xla.core.xla_model as xm
+import torch_xla.runtime as xr
+import torch_xla.experimental.xla_sharding as xs
+from torch_xla.experimental.xla_sharded_tensor import XLAShardedTensor
+
+try:
+  import rich
+  import rich.align
+  import rich.box
+  import rich.console
+  import rich.padding
+  import rich.style
+  import rich.table
+  RICH_ENABLED = True
+except:
+  RICH_ENABLED = False
+
+# Sharding visualization
+sharding_callbacks = weakref.WeakValueDictionary()
+_INSPECT_SHARDING_CALL_NAME = "InspectSharding"
+
+
+class ShardingCallbackInfo:
+
+  def __init__(self, callback, module_context):
+    self.callback = callback
+    self.module_context = module_context
+
+
+Color = Union[tuple[float, float, float], str]
+ColorMap = Callable[[float], tuple[float, float, float, float]]
+
+
+def _canonicalize_color(color: Color) -> str:
+  if isinstance(color, str):
+    return color
+  r, g, b = (int(a * 255) for a in color)
+  return f"#{r:02X}{g:02X}{b:02X}"
+
+
+def _get_text_color(color: str) -> str:
+  r, g, b = torch.map(lambda x: int(x, 16),
+                      (color[1:3], color[3:5], color[5:7]))
+  if (r * 0.299 + g * 0.587 + b * 0.114) > 186:
+    return "#000000"
+  return "#ffffff"
+
+
+def make_color_iter(color_map, num_rows, num_cols):
+  num_colors = num_rows * num_cols
+  color_values = np.linspace(0, 1, num_colors)
+  idx = 0
+  for _ in range(num_colors):
+    yield color_map(color_values[idx])
+    idx = (idx + num_colors // 2 + bool(num_colors % 2 == 0)) % num_colors
+
+
+def visualize_sharding(shape: torch.Size,
+                       sharding: str,
+                       use_color: bool = True,
+                       scale: float = 1.,
+                       min_width: int = 9,
+                       max_width: int = 80,
+                       color_map: Optional[ColorMap] = None):
+  """Visualizes a ``Sharding`` using ``rich``."""
+  if not RICH_ENABLED:
+    raise ValueError("`visualize_sharding` requires `rich` to be installed.")
+
+  if len(shape) > 2 or len(shape) < 1:
```

Comment:
may not need this restriction, based on https://github.com/pytorch/xla/pull/5742/files#r1406976549

Reply:
thanks, updated